addons/mail/models/res_partner.py

```diff
@@ -102,6 +102,97 @@ def find_or_create(self, email, assert_valid_email=False):
             create_values['email'] = parsed_email
         return self.create(create_values)
 
+    @api.model
+    def _find_or_create_from_emails(self, emails, additional_values=None):
+        """ Based on a list of emails, find or create partners. Additional values
+        can be given to newly created partners. If an email is not unique (e.g.
+        multi-email input), only the first found email is considered.
+
+        If no valid email is found for a given item, the given value is used to
+        find partners with same invalid email or create a new one with the wrong
+        value. It allows updating it afterwards. Notably with notifications
+        resend it is possible to update emails, if only a typo prevents from
+        having a real email for example.
+
+        :param list emails: list of emails that may be formatted (each input
+          will be parsed and normalized);
+        :param dict additional_values: additional values given to create if
+          the partner is not found. Typically used to propagate a company_id;
+
+        :return: res.partner records in a list, following order of emails. It
+          is not a recordset, to keep Falsy values.
+        """
+        partners, tocreate_vals_list = self.env['res.partner'], []
+        name_emails = [self._parse_partner_name(email) for email in emails]
+
+        # find valid emails_normalized, filtering out false / void values, and search
+        # for existing partners based on those emails
+        emails_normalized = {email_normalized
+                             for _name, email_normalized in name_emails
+                             if email_normalized}
+        # find partners for invalid (but not void) emails, aka either invalid email
+        # either no email and a name that will be used as email
+        names = {
+            name.strip()
+            for name, email_normalized in name_emails
+            if not email_normalized and name.strip()
+        }
+        if emails_normalized or names:
+            domains = []
+            if emails_normalized:
+                domains.append([('email_normalized', 'in', list(emails_normalized))])
+            if names:
+                domains.append([('email', 'in', list(names))])
+            partners += self.search(expression.OR(domains))
```

Comment:
Hello @xavierbol @tde-banana-odoo ,

You only take into account companies when creating new partners (from additional_values), but here you allow partners from any company to be returned (self.env.su = True from the compute_sudo when you get here).

This can lead to an issue when you have partners with the same email in different companies, eventually returning a partner with a company that the user is not allowed to use from ir.rule: "res.partner company".